File: logic/src/main/java/com/tp/holdem/logic/model/Card.java
package com.tp.holdem.logic.model;

import com.tp.holdem.model.common.Honour;
import com.tp.holdem.model.common.Suit;
import com.tp.holdem.model.message.dto.CardDTO;
import lombok.Builder;
import lombok.Value;

@Value
public class Card {
	private Suit suit;
	private Honour honour;
	private int value;

	@Builder
	private Card(final Suit suit, final Honour honour) {
		this.suit = suit;
		this.honour = honour;
		this.value = this.honour.value();
	}

	public static Card from(Suit suit, Honour honour) {
		return new Card(suit, honour);
	}

	public static Card coded(String code) {
		return Card.from(Suit.coded(code.charAt(code.length() - 1)), Honour.coded(code.substring(0, code.length() - 1)));
	}

	public CardDTO toDTO() {
		return CardDTO.builder()
				.honour(honour)
				.suit(suit)
				.build();
	}
}

File: logic/src/main/kotlin/com/tp/holdem/logic/model/Card.kt
package com.tp.holdem.logic.model

import com.tp.holdem.model.common.Honour
import com.tp.holdem.model.common.Suit
import com.tp.holdem.model.message.dto.CardDTO
import lombok.Builder
import lombok.Value


data class Card(
        val suit: Suit,
        val honour: Honour,
        val value: Int = honour.value()
) {
    companion object {
        @JvmStatic
        fun from(suit: Suit, honour: Honour): Card {
            return Card(suit, honour)
        }

        @JvmStatic
        fun coded(code: String): Card {
            return Card.from(Suit.coded(code[code.length - 1]), Honour.coded(code.substring(0, code.length - 1)))
        }
    }

    fun toDTO(): CardDTO {
        return CardDTO.builder()
                .honour(honour)
                .suit(suit)
                .build()
    }
}